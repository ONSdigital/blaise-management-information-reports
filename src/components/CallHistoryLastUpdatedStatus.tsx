import React, {ReactElement, useEffect, useState} from "react";
import TimeAgo from "react-timeago";
import dateFormatter from "dayjs";
import {getInterviewerCallHistoryStatus} from "../utilities/HTTP";
import {CallHistoryStatus} from "../interfaces";

const CallHistoryLastUpdatedStatus = (): ReactElement => {
    const [reportStatus, setReportStatus] = useState<Date | "">("");
    const [reportStatusFailed, setReportStatusFailed] = useState<boolean>(false);

    useEffect(() => {
        getInterviewerCallHistoryStatus().then((callHistoryStatus: CallHistoryStatus | undefined) => {
            if (!callHistoryStatus) {
                setReportStatusFailed(true);
                return;
            }
            setReportStatus(new Date(callHistoryStatus.last_updated));
        });
    }, []);

    const ReportStatusText = () => {
        if (reportStatusFailed) {
            return (<>Unknown</>);
        }
        return (
            <>
                {<TimeAgo live={false} date={reportStatus}/>}
                {(reportStatus ? "" + dateFormatter(reportStatus).tz("Europe/London").format(" (DD/MM/YYYY HH:mm:ss)") : "Loading")}
            </>
        );
    };

    return (
        <>
            <p className="u-fs-s u-mt-s" aria-live="polite">
                Data in this report was last updated: <b>
                <ReportStatusText/>
            </b>
            </p>
            <p className="u-fs-s">
                Data in this report only goes back to the last 12 months.
            </p>
        </>
    );
};

export default CallHistoryLastUpdatedStatus;
