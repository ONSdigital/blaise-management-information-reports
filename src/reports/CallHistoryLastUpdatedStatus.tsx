import React, {ReactElement, useEffect, useState} from "react";
import TimeAgo from "react-timeago";
import dateFormatter from "dayjs";
import {getInterviewerCallHistoryStatus} from "../utilities/http";

const CallHistoryLastUpdatedStatus = (): ReactElement => {
    const [reportStatus, setReportStatus] = useState<Date | "">("");
    const [reportStatusFailed, setReportStatusFailed] = useState<boolean>(false);

    useEffect(() => {
        getInterviewerCallHistoryStatus().then(([success, last_updated]) => {
            if (!success) {
                setReportStatusFailed(true);
            }
            setReportStatus(new Date(last_updated.last_updated));
        });
    }, []);

    const ReportStatusText = () => {
        if (reportStatusFailed) {
            return (<>Unknown</>);
        }
        return (
            <>
                {<TimeAgo live={false} date={reportStatus}/>}
                {(reportStatus ? "" + dateFormatter(reportStatus).tz("Europe/London").format(" (DD/MM/YYYY HH:mm:ss)") : "Loading")}
            </>
        );
    };

    return (
        <>
            <p className="u-fs-s u-mt-s" aria-live="polite">
                Data in this report was last updated: <b>
                <ReportStatusText/>
            </b>
            </p>
            <p className="u-fs-s">
                Data in this report only goes back to the last 12 months.
            </p>
        </>
    );
};

export default CallHistoryLastUpdatedStatus;
