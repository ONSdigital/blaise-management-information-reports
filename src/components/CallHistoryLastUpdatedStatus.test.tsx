/**
 * @jest-environment jsdom
 */

import "@testing-library/jest-dom";
import flushPromises from "../tests/utilities";
import {createMemoryHistory} from "history";
import {cleanup, render, waitFor} from "@testing-library/react";
import {Router} from "react-router";
import CallHistoryLastUpdatedStatus from "./CallHistoryLastUpdatedStatus";
import {act} from "react-dom/test-utils";
import {screen} from "@testing-library/dom";
import React from "react";
import dateFormatter from "dayjs";
import utc from "dayjs/plugin/utc";
import timezone from "dayjs/plugin/timezone";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";

const mockAdapter = new MockAdapter(axios);

dateFormatter.extend(utc);
dateFormatter.extend(timezone);

function subtractYears(numOfYears: any, date: Date = new Date()) {
    date.setFullYear(date.getFullYear() - numOfYears);
    return date;
}

const time = subtractYears(1)

describe("call history last updated status with data", () => {
    beforeEach(() => {
        mockAdapter.reset();

        mockAdapter.onGet("/api/reports/call-history-status").reply(200,
            {"last_updated": time});
    });

    it("matches snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <CallHistoryLastUpdatedStatus/>
            </Router>
        );
        await act(async () => {
            await flushPromises();
        });
        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("renders correctly", async () => {
        const history = createMemoryHistory();
        await act(async () => {
            render(
                <Router history={history}>
                    <CallHistoryLastUpdatedStatus/>
                </Router>
            );
        });
        expect(screen.queryByText("Data in this report was last updated:")).toBeVisible();
        expect(screen.queryByText("Data in this report only goes back to the last 12 months.")).toBeVisible();
        await act(async () => {
            await flushPromises();
        });
        await waitFor(() => {
            expect(screen.getByText(dateFormatter(time).format("(DD/MM/YYYY HH:mm:ss)"))).toBeVisible();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("call history last updated status with invalid data", () => {
    beforeEach(() => {
        mockAdapter.reset();

        mockAdapter.onGet("/api/reports/call-history-status").reply(200, "blah");
    });

    it("matches snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <CallHistoryLastUpdatedStatus/>
            </Router>
        );
        await act(async () => {
            await flushPromises();
        });
        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("renders correctly", async () => {
        const history = createMemoryHistory();
        await act(async () => {
            render(
                <Router history={history}>
                    <CallHistoryLastUpdatedStatus/>
                </Router>
            );
        });
        expect(screen.queryByText("Data in this report was last updated:")).toBeVisible();
        expect(screen.queryByText("Data in this report only goes back to the last 12 months.")).toBeVisible();
        await act(async () => {
            await flushPromises();
        });
        await waitFor(() => {
            expect(screen.getByText("Invalid Date")).toBeVisible();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
