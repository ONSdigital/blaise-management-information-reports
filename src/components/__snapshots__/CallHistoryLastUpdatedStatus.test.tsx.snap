// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call history last updated status with data matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        aria-live="polite"
        class="u-fs-s u-mt-s"
      >
        Data in this report was last updated: 
        <b>
          <time
            datetime="2000-01-01T10:00:00.000Z"
            title="2000-01-01 10:00"
          >
            22 years ago
          </time>
           (01/01/2000 10:00:00)
        </b>
      </p>
      <p
        class="u-fs-s"
      >
        Data in this report only goes back to the last 12 months.
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      aria-live="polite"
      class="u-fs-s u-mt-s"
    >
      Data in this report was last updated: 
      <b>
        <time
          datetime="2000-01-01T10:00:00.000Z"
          title="2000-01-01 10:00"
        >
          22 years ago
        </time>
         (01/01/2000 10:00:00)
      </b>
    </p>
    <p
      class="u-fs-s"
    >
      Data in this report only goes back to the last 12 months.
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`call history last updated status with invalid data matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        aria-live="polite"
        class="u-fs-s u-mt-s"
      >
        Data in this report was last updated: 
        <b>
          Invalid Date
        </b>
      </p>
      <p
        class="u-fs-s"
      >
        Data in this report only goes back to the last 12 months.
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      aria-live="polite"
      class="u-fs-s u-mt-s"
    >
      Data in this report was last updated: 
      <b>
        Invalid Date
      </b>
    </p>
    <p
      class="u-fs-s"
    >
      Data in this report only goes back to the last 12 months.
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
